prusti-specs/src/lib.rs

```diff
@@ -21,34 +26,110 @@ macro_rules! handle_result {
+    prusti_attributes.extend(
+        remaining_prusti_attributes.map(
+            |attr| attr.path.segments[0].ident.to_string().try_into().ok().map(
+                |attr_kind| (attr_kind, attr.tokens)
+            )
+        ).flatten()
+    );
+
+    let (gen_spec_items, gen_attributes) = generate_spec_and_assertions(prusti_attributes, &item);
+
+    quote!{
+        #(#gen_spec_items)*
+        #(#gen_attributes)*
+        #item
+    }
+}
+
+/// Generate spec items and item attributes from the Prusti attributes
+fn generate_spec_and_assertions(
+    mut prusti_attributes: Vec<(SpecAttributeKind, TokenStream)>,
+    item: syn::ItemFn
+) -> RewritingResult {
+    let mut generated_items = vec![];
+    let mut generated_attributes = vec![];
+
+    for (attr_kind, attr_tokens) in prusti_attributes.drain(..) {
+        let rewriting_result = match attr_kind {
+            SpecAttributeKind::Requires => generate_for_requires(attr_tokens, &item),
+            SpecAttributeKind::Ensures => generate_for_ensures(attr_tokens, &item),
+            SpecAttributeKind::AfterExpiry => generate_for_after_expiry(attr_tokens, &item),
+            SpecAttributeKind::AfterExpiryIf => generate_for_after_expiry_if(attr_tokens, &item),
+            SpecAttributeKind::Pure => generate_for_pure(attr_tokens, &item),
+            SpecAttributeKind::Trusted => generate_for_trusted(attr_tokens, &item),
+        };
+        let (new_items, new_attributes) = handle_result!(rewriting_result);
+        generated_items.extend(new_items);
+        generated_attributes.extend(new_attributes);
+    }
+
+    quote!{
+        #(#generated_items)*
+        #(#generated_attributes)*
+        #item
+    }
+}
+
+type RewritingResult = syn::Result<(Vec<syn::Item>, Vec<syn::Attribute>)>;
+
+/// As required by the "requires" annotation, generate spec items and attributes.
+fn generate_for_requires(attr: TokenStream, item: &syn::ItemFn) -> RewritingResult {
     let mut rewriter = rewriter::AstRewriter::new();
     let spec_id = rewriter.generate_spec_id();
     let spec_id_str = spec_id.to_string();
-    let assertion = handle_result!(rewriter.parse_assertion(spec_id, attr));
-    let spec_item =
-        handle_result!(rewriter.generate_spec_item_fn(rewriter::SpecItemType::Precondition, spec_id, assertion, &item));
-    quote! {
-        #spec_item
-        #[prusti::pre_spec_id_ref = #spec_id_str]
-        #item
-    }
+    let assertion = rewriter.parse_assertion(spec_id, attr)?;
+    let spec_item = rewriter.generate_spec_item_fn(
+        rewriter::SpecItemType::Precondition,
+        spec_id,
+        assertion,
+        &item
+    )?;
+    Ok((
+        vec![spec_item],
+        vec![parse_quote!(#[prusti::pre_spec_id_ref = #spec_id_str])],
+    ))
 }
 
-pub fn ensures(attr: TokenStream, tokens: TokenStream) -> TokenStream {
-    let item: syn::ItemFn = handle_result!(syn::parse2(tokens));
+/// As required by the "ensures" annotation, generate spec items and attributes.
```

Comment:
I'm not sure how useful these comments are.

Reply:
I'm too not happy about them

Reply:
It should be slightly better now.